File: cSharp/easypay/EasypayRequest.cs
namespace easypay.swisscom.com
{
    using System;
    using System.Globalization;
    using System.Net;
    using System.Net.Http;
    using System.Net.Http.Headers;
    using System.Text;
    using System.Threading.Tasks;
    using Newtonsoft.Json;
    using Newtonsoft.Json.Serialization;

    public class EasypayRequest
    {
        public static Uri GetCheckoutPageUrl(EasypayConfig config, CheckoutPageRequest paymentRequest)
        {
            var data = JsonConvert.SerializeObject(paymentRequest, new JsonSerializerSettings { ContractResolver = new CamelCasePropertyNamesContractResolver() });
            var dataByteArray = Encoding.UTF8.GetBytes(data);
            var signature = Signature.Sign(Encoding.UTF8.GetBytes(config.EasypaySecret), dataByteArray);
            var path = "/authorize.jsf?signature=" + Uri.EscapeDataString(Convert.ToBase64String(signature)) + "&checkoutRequestItem=" + Uri.EscapeDataString(Convert.ToBase64String(dataByteArray));
            return new Uri("http://" + config.Host + config.Basepath + path);
        }

        public static async Task<PaymentResponse> CommitPaymentRequest(string paymentId, EasypayConfig easypayConfig, string reqId)
        {
            var dataJSON = @"{""operation"":""COMMIT""}";
            var path = "/payments/" + paymentId;
            return await ProcessPaymentRequest(dataJSON, path, easypayConfig, reqId);
        }

        public static async Task<PaymentResponse> CancelPaymentRequest(string paymentId, EasypayConfig easypayConfig, string reqId)
        {
            var dataJSON = @"{""operation"":""CANCEL""}";
            var path = "/payments/" + paymentId;
            return await ProcessPaymentRequest(dataJSON, path, easypayConfig, reqId);
        }

        public static async Task<PaymentResponse> ProcessPaymentRequest(string dataJson, string path, EasypayConfig easypayConfig, string reqId)
            { 
            var resultMessage = await ProcessRequest(easypayConfig, dataJson, HttpMethod.Put, path, reqId);
            Console.WriteLine("Response: " + resultMessage.Response);

            var ret = JsonConvert.DeserializeObject<PaymentResponse>(resultMessage.Response) ?? new PaymentResponse();
            ret.Success = resultMessage.StatusCode == HttpStatusCode.OK;
            return ret;
        }

        public static async Task<StatusResponse> GetPaymentStatus(string paymentId, EasypayConfig easypayConfig, string reqId)
        {
            var resultMessage = await ProcessRequest(easypayConfig, null, HttpMethod.Get, "/payments/" + paymentId, reqId);
            Console.WriteLine("Response: " + resultMessage.Response);

            var ret = JsonConvert.DeserializeObject<StatusResponse>(resultMessage.Response) ?? new StatusResponse();
            return ret;
        }

        private static async Task<RequestResponse> ProcessRequest(EasypayConfig config, string data, HttpMethod httpMethod, string path, string reqId)
        {
            var contentType = httpMethod.Method == "GET" ? "" : "application/vnd.ch.swisscom.easypay.direct.payment+json";
            var url = new Uri("https://" + config.Host + config.Basepath + path);
            var now = DateTime.Now;
            var date = now.ToUniversalTime().ToString("ddd, dd MMM yyyy HH:mm:ss", CultureInfo.GetCultureInfoByIetfLanguageTag("en")) + " +0000"; //Mon, 07 Dec 2015 09:01:30 +0000

            var client = new HttpClient();
            var request = new HttpRequestMessage(httpMethod, url);
            request.Headers.Add("X-SCS-Date", date);
            request.Headers.Add("X-Request-Id", reqId);
            request.Headers.Add("X-Merchant-Id", config.MerchantId);
            request.Headers.Add("X-CE-Client-Specification-Version", "1.1");

            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/vnd.ch.swisscom.easypay.message.list+json"));
            request.Headers.Date = now;

            var md5Hash = data != null ? Signature.HashData(Encoding.UTF8.GetBytes(data)) : null;
            var hashString = Signature.CreateHashString(httpMethod.Method, md5Hash != null ? Convert.ToBase64String(md5Hash) : "", contentType, date, path);
            var signature = Signature.Sign(Encoding.UTF8.GetBytes(config.EasypaySecret), Encoding.UTF8.GetBytes(hashString));
            
            request.Headers.Add("X-SCS-Signature", Convert.ToBase64String(signature));

            if (data != null)
            {
                request.Content.Headers.ContentType = new MediaTypeHeaderValue(contentType);
                request.Content = new StringContent(data);
                request.Content.Headers.ContentMD5 = md5Hash;
            }

            var result = await client.SendAsync(request);
            var ret = new RequestResponse {Response = await result.Content.ReadAsStringAsync(), StatusCode = result.StatusCode};
            return ret;
        }

        private class RequestResponse
        {
            public string Response { get; set; }
            public HttpStatusCode StatusCode { get; set; }
        }
    }
}

File: javascript/easpay.lib/EasypayRequest.ts
import moment = require('moment');
import https = require('https');
import Promise = require('bluebird');

import SignatureNS = require("./Signature");
import PaymentResponseNS = require("./IPaymentResponse");
import StatusResponseNS = require("./IStatusResponse");
var Signature = SignatureNS.easypay.Signature;

export module easypay {
    export module EasypayRequest {

        export function getCheckoutPageUrl(config: IEasypayConfig, paymentRequest: any): string {
            var data = JSON.stringify(paymentRequest);
            var encodedData = new Buffer(data).toString('base64');
            var signature = Signature.sign(config.easypaySecret, data);
            var path = "/authorize.jsf?signature=" + encodeURIComponent(signature) + "&checkoutRequestItem=" + encodeURIComponent(encodedData);
            return `${config.protocol}://${config.host}${config.basepath}${path}`;
        }

        export function commitPaymentRequest(paymentId: string, easypayConfig: IEasypayConfig, reqId: string): Promise<PaymentResponseNS.IPaymentResponse> {
            return processRequest(easypayConfig, JSON.stringify({ operation: "COMMIT" }), "PUT", `/payments/${paymentId}`, reqId);
        }

        export function cancelPaymentRequest(paymentId: string, easypayConfig: IEasypayConfig, reqId: string): Promise<PaymentResponseNS.IPaymentResponse> {
            return processRequest(easypayConfig, JSON.stringify({ operation: "CANCEL" }), "PUT", `/payments/${paymentId}`, reqId);
        }

        export function getPaymentStatus(paymentId, easypayConfig, reqId) :Promise<StatusResponseNS.IStatusResponse> {
            return processRequest(easypayConfig, null, "GET", "/payments/" + paymentId, reqId);
        }

        function processRequest(config: IEasypayConfig, data: string, httpMethod: string, path: string, reqId: string): any {
            var date = moment().utc().format("ddd, DD MMM YYYY HH:mm:ss ZZ");
            var contentType = httpMethod === "GET" ? "" : "application/vnd.ch.swisscom.easypay.direct.payment+json";

            var options = {
                host: config.host,
                path: config.basepath + path,
                method: httpMethod,
                headers: {
                    "X-SCS-Date": date,
                    "X-Request-Id": reqId || new Date().getTime(),
                    "X-Merchant-Id": config.merchantId,
                    "Accept": "application/vnd.ch.swisscom.easypay.message.list+json",
                    "X-CE-Client-Specification-Version": "1.1",
                    "Date": date
                }
            };

            var md5Hash = "";

            if (data) {
                md5Hash = Signature.hashData(data);
                options.headers["Content-MD5"] = md5Hash;
            }

            var hashString = Signature.createHashString(httpMethod, md5Hash, contentType, date, path);
            var signature = Signature.sign(config.easypaySecret, hashString);

            if (contentType) {
                options.headers["Content-Type"] = contentType;
            }

            options.headers["X-SCS-Signature"] = signature;

            return request(options, data);
        }

        function request(options, requestData): Promise<PaymentResponseNS.IPaymentResponse>{

            return new Promise<PaymentResponseNS.IPaymentResponse>((resolve, reject) => {
                var ret = <PaymentResponseNS.IPaymentResponse>{ success: false };

                try {
                    var request = https.request(options, (result) => {
                        result.setEncoding("utf8");
                        var data = "";

                        result.on("data", (chunk) => {
                            data += chunk;
                        });

                        result.on("end", () => {
                            console.log(data);
                            if (result.statusCode !== 200) {
                                try {
                                    ret.messages = JSON.parse(data)['messages'];
                                } catch (e) {
                                    ret.messages = [];
                                    ret.error = data;
                                }
                                reject(ret);
                            } else {
                                if (data)
                                    resolve(JSON.parse(data) || { success: true });
                                else
                                    resolve(<PaymentResponseNS.IPaymentResponse>{ success: true });
                            }
                        });
                    });

                    // Handle errors
                    request.on("error", (error) => {
                        ret.error = error;
                        reject(ret);
                    });

                    // send the data
                    if (requestData) {
                        request.write(requestData);
                    }

                    request.end();
                } catch (e) {
                    ret.error = e;
                    reject(ret);
                }
            });
        }
    }
}
